src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,469 @@
+    }
+    uint256 constant RAY  = 10 ** 27;
+    function _rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, y) / RAY;
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Administration ---
+    function file(bytes32 ilk_, bytes32 what, uint256 data) external auth {
+        Ilk storage ilk = ilks[ilk_];
+        if (what == "bar") {
+            require(data <= ilk.join.getMaxBar(), "DssDirectDepositHub/above-max-interest");
+
+            ilk.bar = data;
+        } else if (what == "tau" ) {
+            require(live == 1, "DssDirectDepositHub/hub-not-live");
+            require(ilk.tic == 0, "DssDirectDepositHub/join-not-live");
+
+            ilk.tau = data;
+        } else revert("DssDirectDepositHub/file-unrecognized-param");
+
+        emit File(ilk_, what, data);
+    }
+
+    function file(bytes32 ilk, bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+        require(ilks[ilk].tic == 0, "DssDirectDepositHub/join-not-live");
+
+        if (what == "king") ilks[ilk].king = data;
+        else if (what == "join") ilks[ilk].join = DssDirectDepositJoinLike(data);
+        else if (what == "gem") ilks[ilk].gem = TokenLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+
+    // --- Deposit controls ---
+    function _wind(bytes32 ilk, DssDirectDepositJoinLike join, uint256 amount) internal {
+        // IMPORTANT: this function assumes Vat rate of this ilk will always be == 1 * RAY (no fees).
+        // That's why this module converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+
+        // Wind amount is limited by the debt ceiling
+        (uint256 Art,,, uint256 line,) = vat.ilks(ilk);
+        uint256 lineWad = line / RAY; // Round down to always be under the actual limit
+        if (_add(Art, amount) > lineWad) {
+            amount = _sub(lineWad, Art);
+        }
+
+        if (amount == 0) {
+            emit Wind(ilk, 0);
+            return;
+        }
+
+        require(int256(amount) >= 0, "DssDirectDepositHub/overflow");
+
+        uint256 scaledPrev = join.getNormalizedBalanceOf();
+
+        vat.slip(ilk, address(join), int256(amount));
+        vat.frob(ilk, address(join), address(join), address(join), int256(amount), int256(amount));
+        // normalized debt == erc20 DAI (Vat rate for this ilk fixed to 1 RAY)
+        join.supply(amount);
+
+        // Verify the correct amount of gem shows up
+        uint256 scaledAmount = join.getNormalizedAmount(amount);
```

Comment:
I think the scaling/normalization part of the gem balance should be abstracted away in this contract in case other future collaterals work in slightly different ways. Or maybe this whole check should be done inside of `join.supply()` ?

```suggestion
        uint256 nextGemBalance = join.getGemBalance(amount);
```